src/sparse/impl/KokkosSparse_spgemm_mkl_impl.hpp

```diff
@@ -47,634 +47,383 @@
-#include <Kokkos_Concepts.hpp>
-
 namespace KokkosSparse {
-
 namespace Impl {
 
-template <typename KernelHandle, typename in_row_index_view_type,
-          typename in_nonzero_index_view_type, typename bin_row_index_view_type,
-          typename bin_nonzero_index_view_type,
-          typename cin_row_index_view_type>
-void mkl_symbolic(KernelHandle *handle, typename KernelHandle::nnz_lno_t m,
-                  typename KernelHandle::nnz_lno_t n,
-                  typename KernelHandle::nnz_lno_t k,
-                  in_row_index_view_type row_mapA,
-                  in_nonzero_index_view_type entriesA,
-
-                  bool transposeA, bin_row_index_view_type row_mapB,
-                  bin_nonzero_index_view_type entriesB, bool transposeB,
-                  cin_row_index_view_type row_mapC, bool verbose = false) {
 #ifdef KOKKOSKERNELS_ENABLE_TPL_MKL
 
-  typedef typename KernelHandle::nnz_lno_t idx;
-  typedef typename KernelHandle::size_type size_type;
-
-  typedef typename KernelHandle::HandleTempMemorySpace HandleTempMemorySpace;
-  typedef typename Kokkos::View<int *, HandleTempMemorySpace>
-      int_temp_work_view_t;
+inline void mkl_call(sparse_status_t result, const char *err_msg) {
+  if (SPARSE_STATUS_SUCCESS != result) {
+    throw std::runtime_error(err_msg);
+  }
+}
 
-  typedef typename KernelHandle::nnz_scalar_t value_type;
+template <typename value_type>
+class MKLSparseMatrix {
+  sparse_matrix_t mtx;
+
+ public:
+  inline MKLSparseMatrix(const MKL_INT m, const MKL_INT n, MKL_INT *xadj,
+                         MKL_INT *adj, value_type *values);
+
+  inline static MKLSparseMatrix<value_type> spmm(
+      sparse_operation_t operation, const MKLSparseMatrix<value_type> &A,
+      const MKLSparseMatrix<value_type> &B) {
+    sparse_matrix_t c;
+    mkl_call(mkl_sparse_spmm(operation, A.mtx, B.mtx, &c),
+             "mkl_sparse_spmm() failed!");
+    return MKLSparseMatrix<value_type>(c);
+  }
 
-  typedef typename KernelHandle::HandleExecSpace MyExecSpace;
-  /*
-    if (!(
-        (Kokkos::SpaceAccessibility<typename
-    Kokkos::HostSpace::execution_space, typename
-    device1::memory_space>::accessible) &&
-        (Kokkos::SpaceAccessibility<typename
-    Kokkos::HostSpace::execution_space, typename
-    device2::memory_space>::accessible) &&
-        (Kokkos::SpaceAccessibility<typename
-    Kokkos::HostSpace::execution_space, typename
-    device3::memory_space>::accessible) )
-        ){
-      throw std::runtime_error ("MEMORY IS NOT ALLOCATED IN HOST DEVICE for
-    MKL\n"); return;
-    }
-  */
-  if (std::is_same<idx, int>::value) {
-    int *a_xadj = NULL;
-    int *b_xadj = NULL;
-    int_temp_work_view_t a_xadj_v, b_xadj_v;
-
-    if (std::is_same<size_type, int>::value) {
-      a_xadj = (int *)row_mapA.data();
-      b_xadj = (int *)row_mapB.data();
-    } else {
-      // TODO test this case.
+  inline void get(MKL_INT &rows, MKL_INT &cols, MKL_INT *&rows_start,
```

Comment:
The function name is not very explicit, what are you getting with `get`?
Also can we reuse names similar to those used in `CrsMatrix`: `numrows`, `numcols`... for simplicity?

Reply:
It retrieves all matrix information. Would `export_data()` sound better ?  
For instance:
```cpp
Matrix C = spmm(operation, A, B);
...
C.export_data(num_rows, num_cols, rows_start, columns, values);
```